schemas/irc-request.yml

```diff
@@ -1,33 +1,44 @@
 $schema:            http://json-schema.org/draft-06/schema#
 title:              "Post IRC Message Request"
-description: |
-  Request to post a message on IRC.
-type:               object
-properties:
-  channel:
-    type:           string
-    pattern:        "^[#&][^ ,\\u0007]{1,199}$"
-    description: |
-      Channel to post the message in. Please note that you **must** supply
-      either `user` or `channel`, you cannot supply both.
-  user:
-    type:           string
-    pattern:        "^[A-Za-z\\[\\]\\\\~_\\^{|}][A-Za-z0-9\\-\\[\\]\\\\~_\\^{|}]{0,254}$"
-    maxLength:      255
-    minLength:      1
-    description: |
-      User to post the message to. Please note that you **must** supply
-      either `user` or `channel`, you cannot supply both.
+definitions:
   message:
+    title:          IRC Message Text
     type:           string
     minLength:      1
     maxLength:      510
     description: |
       IRC message to send as plain text.
-maxProperties: 2
-additionalProperties: false
```

Comment:
@petemoore, how about tweaking it to:
```yaml
maxProperties: 2
minProperties: 2
additionalProperties: false
required:
  - message
```

that way we remove the `oneOf` and everybody wins :)

Reply:
I don't understand, sorry!

Reply:
```yaml
$schema:            http://json-schema.org/draft-06/schema#
title:              "Post IRC Message Request"
description: |
  Request to post a message on IRC.
type:               object
properties:
  channel:
    type:           string
    pattern:        "^[#&][^ ,\\u0007]{1,199}$"
    description: |
      Channel to post the message in. Please note that you **must** supply
      either `user` or `channel`, you cannot supply both.
  user:
    type:           string
    pattern:        "^[A-Za-z\\[\\]\\\\~_\\^{|}][A-Za-z0-9\\-\\[\\]\\\\~_\\^{|}]{0,254}$"
    maxLength:      255
    minLength:      1
    description: |
      User to post the message to. Please note that you **must** supply
      either `user` or `channel`, you cannot supply both.
  message:
    type:           string
    minLength:      1
    maxLength:      510
    description: |
      IRC message to send as plain text.
maxProperties: 2
minProperties: 2  # <--- I added this, so we don't need the oneOf
additionalProperties: false
required:
  - message
# I removed this:
#oneOf:
#  - required: [channel]
#  - required: [user]
```

Okay, what I wrote below got long, **feel free to ignore it** :)

I only wanted to suggest the schema above instead as this would completely avoid the `oneOf`.
It might be slightly worse for documentation, but no worse than what we have now. And it'll be much nicer in the generated types.

-----

I think this schema is another instance of the case where there is enough type information available to generate the type without looking at `oneOf/allOf/anyOf/not`.

Using this trick we can work around using `oneOf`, but in this case it's not here to communicate type information, that's communicated higher up where we have `type: object; properties: ...`.
The `oneOf` is only here to facilitate extra validation.

note: I agree we should strive to reflect the schema in the type. Including `description`, `pattern` and such is super useful. But there is limits, not all data validation is expressed in the schema, validation also happens in the request handler, and if not all validation information from the schema is available in the type documentation that's also acceptable.

For example: I don't expect `not` to be represented in the types generated from the schema. `not` is useful as documentation and validation feature, but it could just as well be a hidden validation that happens inside the API handler serverside.